opm/common/ConditionalStorage.hpp

```diff
@@ -0,0 +1,167 @@
+// -*- mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+// vi: set et ts=4 sw=4 sts=4:
+/*
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 2 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+
+  Consult the COPYING file in the top-level source directory of this
+  module for the precise wording of the license and the list of
+  copyright holders.
+*/
+/*!
+ * \file
+ *
+ * \copydoc Opm::ConditionalStorage
+ */
+#ifndef OPM_CONDITIONAL_STORAGE_HH
+#define OPM_CONDITIONAL_STORAGE_HH
+
+#include <opm/common/ErrorMacros.hpp>
+#include <opm/common/Exceptions.hpp>
+#include <opm/common/Unused.hpp>
+
+#include <utility>
+
+namespace Opm {
+/*!
+ * \ingroup Common
+ *
+ * \brief A simple class which only stores a given member attribute if a boolean
+ *        condition is true
+ *
+ * If the condition is false, nothing is stored and an exception is thrown when trying to
+ * access the object.
+ */
+template <bool cond, class T>
+class ConditionalStorage
+{
+public:
+    typedef T type;
+    static constexpr bool condition = cond;
+
+    ConditionalStorage()
+    {}
+
+    ConditionalStorage(const T& v)
+        : data_(v)
+    {}
+
+    ConditionalStorage(T&& v)
+        : data_(v)
```

Comment:
Don't you need an std::move here to actually get the behaviour you want?

Reply:
oops. you're right. the move mechanism of C++ is IMO way too  easy to get wrong.

changed & pushed, thanks.